### PR TITLE
Take RRs more into account for unification

### PR DESCRIPTION
**Pull request by yannl35133**
Not 100% sure that this is correct, but this fixes an issue (using code from another branch so I can't put it in the test-suite)

### CODE REVIEW

**Comment by SkySkimmer** [comment]
"Take RRs more into account for unification and compare projections by repr name" sounds like it should be 2 commits (splitting at the `and`)

**Comment by yannl35133** [comment]
The projection repr part is now split off in #19509

File: pretyping/reductionops.ml [hunk 1]
```diff
@@ -1671,6 +1671,16 @@ let whd_betaiota_deltazeta_for_iota_state ts env sigma s =
...
+    match match kind sigma t with
+      | Const (cst, u) when Environ.is_symbol env cst ->
+        begin
+          try
+            let r = match Cmap_env.find_opt cst env.symb_pats with Some r -> r | None -> assert false in
+            Some (apply_rules whrec env sigma u r stack)
+          with PatternFailure -> None
+        end
+      | _ -> None
+    with Some r -> whrec r | None ->
```
Comment (SkySkimmer) [inline pretyping/reductionops.ml:1683]:
```suggestion
    let rewrite_step = match kind sigma t with
      | Const (cst, u) when Environ.is_symbol env cst ->
        begin
          try
            let r = match Cmap_env.find_opt cst env.symb_pats with Some r -> r | None -> assert false in
            Some (apply_rules whrec env sigma u r stack)
          with PatternFailure -> None
        end
      | _ -> None
    in
    match rewrite_step with 
    | Some r -> whrec r
    | None ->
```

File: pretyping/reductionops.ml [hunk 2]
```diff
@@ -1671,6 +1671,16 @@ let whd_betaiota_deltazeta_for_iota_state ts env sigma s =
   in
   let rec whrec s =
     let (t, stack as s) = whd_state_gen RedFlags.betaiota env sigma s in
+    match match kind sigma t with
+      | Const (cst, u) when Environ.is_symbol env cst ->
+        begin
+          try
+            let r = match Cmap_env.find_opt cst env.symb_pats with Some r -> r | None -> assert false in
```
Comment (SkySkimmer) [inline pretyping/reductionops.ml:1678]:
```suggestion
            let r = Cmap_env.get cst env.symb_pats in
```

**Review by SkySkimmer — APPROVED** [review]

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Pull request merged**